Simplenote/src/androidTest/java/com/automattic/simplenote/integration/TagDialogFragmentTest.kt

```diff
@@ -0,0 +1,326 @@
+        val note2 = createNote("Hello1", listOf("tag2", "tag3"))
+        val note3 = createNote("Hello1", listOf("tag1", "tag3"))
+
+        assertEquals(tagsBucket.count(), 3)
+        assertEquals(notesBucket.count(), 3)
+
+        launchFragment("tag2")
+
+        onView(allOf(withText("tag2"), isDescendantOfA(withId(R.id.input_tag_name)))).check(matches(isDisplayed()))
+
+        onView(allOf(withText("tag2"), isDescendantOfA(withId(R.id.input_tag_name))))
+                .perform(ViewActions.replaceText("tag1"))
+
+        onView(allOf(withText("tag1"), isDescendantOfA(withId(R.id.input_tag_name)))).check(matches(isDisplayed()))
+
+        val saveText = getResourceString(R.string.save)
+        onView(withText(saveText)).inRoot(isDialog()).check(matches(isDisplayed())).perform(click())
+
+        val canonical = TagUtils.getCanonicalFromLexical(tagsBucket, "tag1")
+        val mergeMessage = getResourceStringWithArgs(R.string.dialog_tag_conflict_message, canonical, "tag2", canonical)
+        onView(withId(R.id.message)).check(matches(withText(mergeMessage)))
+
+        val mergeText = getResourceString(R.string.dialog_tag_conflict_button_positive)
+        onView(withText(mergeText)).perform(click())
+
+        val tag1 = getTag("tag1")
+        val tag2 = getTag("tag2")
+        val tag3 = getTag("tag3")
+
+        assertNull(tag2)
+        assertNotNull(tag1)
+        assertNotNull(tag3)
+
+        assertEquals(note1.tags, listOf("tag1"))
+        assertEquals(note2.tags, listOf("tag3", "tag1"))
+        assertEquals(note3.tags, listOf("tag1", "tag3"))
+    }
+
+    @Test
+    fun editTagFails() {
+        createTag("tag1")
+        // To edit tags, tags should belong a note
+        val note = createNote("Hello World", listOf("tag1"))
+
+        assertEquals(tagsBucket.count(), 1)
+        assertEquals(notesBucket.count(), 1)
+
+        launchFragment("tag1")
+
+        onView(allOf(withText("tag1"), isDescendantOfA(withId(R.id.input_tag_name)))).check(matches(isDisplayed()))
+        onView(allOf(withText("tag1"), isDescendantOfA(withId(R.id.input_tag_name))))
+                .perform(ViewActions.replaceText("tag5"))
+        onView(allOf(withText("tag5"), isDescendantOfA(withId(R.id.input_tag_name)))).check(matches(isDisplayed()))
+
+        tagsBucket.newObjectShouldFail = true
+
+        val saveText = getResourceString(R.string.save)
+        onView(withText(saveText)).inRoot(isDialog()).check(matches(isDisplayed())).perform(click())
+
+        val okText = getResourceString(android.R.string.ok)
+        onView(withText(okText)).check(matches(isDisplayed())).perform(click())
+
+        val tag1 = getTag("tag1")
+        val tag5 = getTag("tag5")
+
+        assertNotNull(tag1)
+        assertNull(tag5)
+        assertEquals(tagsBucket.count(), 1)
+        assertEquals(note.tags, listOf("tag1"))
+    }
+
+    private fun launchFragment(tagName: String): FragmentScenario<TagDialogFragment> {
```

Comment:
Question (❓): The return value is not used anywhere, is it? Why would you want to return it anyway, is this something that will be needed later on in subsequent tests?

Reply:
Not in this particular test. I will remove it.

Reply:
Fixed in 801a30df4dd779591070e5eecdf28361191cb66c.